rfc-157-edge-compute.md

```diff
@@ -0,0 +1,236 @@
+
+As the only common language shared between the two platforms, JavaScript (or TypeScript) seems like the obvious choice.
+
+The primary benefit TypeScript provides over JavaScript is type safety, including enforcement of `readonly` properties. This will be particularly useful on the Lambda@Edge side because parts of the payload we receive from CloudFront are intended to be read-only. TypeScript enforces these invariants (potentially preventing us from introducing bugs), whereas JavaScript doesn't know how to.
+
+If folks have a strong aversion to TypeScript, JavaScript would also be a good option.
+
+## Architecture
+
+I propose that we build a three-layer architecture, with:
+
+- A platform-agnostic wrapper API to abstract away differences between the platform-specific types (e.g. `Request` and `Response`)
+- Platform-agnostic request and response handlers (corresponding to Lambda@Edge's viewer request and viewer response handlers concepts respectively), where the majority of the code lives, e.g. (pseudocode)
+  ```typescript
+  // Takes the request from the client
+  // Returns either a request to be sent to the backend,
+  // or a response to be sent to the client
+  function handle_request(request: Request): Request | Response
+
+  // Takes the response returned from either the request handler or the backend
+  // Returns the response to be sent to the client
+  function handle_response(response: Response): Response
+  ```
+- Platform-specific entrypoints for Compute@Edge and Lambda@Edge, which call into the platform-agnostic code, e.g. (pseudocode)
+  ```typescript
+  // Compute@Edge entrypoint:
+  function perform_backend_fetch(request: Request): Response; // Provided by Fastly SDK
+
+  function fastly_entrypoint(request: Request): Response {
+    const request_or_response = handle_request(request)
+    const response = (request_or_response instanceof Request) ?
+      perform_backend_fetch(request_or_response) :
+      request_or_response
+    return handle_response(response)
+  }
+
+  // Lambda@Edge viewer request handler entrypoint:
+  function viewer_request_handler(request: Request): Request | Response {
+    return handle_request(request)
+  }
+
+  // Lambda@Edge viewer response handler entrypoint:
+  function viewer_response_handler(response: Response): Response {
+    return handle_response(response)
+  }
+  ```
+
+On the Fastly side, we will also need:
+
+- A VCL service _before_ our Compute@Edge one so that we can continue to use shielding on the `/alerts` route (and potentially roll out shielding for all of GOV.UK), and
```

Comment:
[Cloudfront has a similar shielding concept](https://docs.aws.amazon.com/AmazonCloudFront/latest/DeveloperGuide/origin-shield.html), do we think this would work OK?